File: web/app/services/appConfigService.ts
import { API_URL } from './api';
import { fetchApi, ApiResponse } from './fetchApi';

// 应用配置接口
export interface AppConfigInput {
  appId: string;
  name: string;
  organizationName: string;
  repositoryName: string;
  allowedDomains: string[];
  enableDomainValidation: boolean;
  description: string;
}

export interface AppConfigOutput {
  appId: string;
  name: string;
  organizationName: string;
  repositoryName: string;
  allowedDomains: string[];
  enableDomainValidation: boolean;
  description: string;
  createdAt: string;
  updatedAt: string;
  isEnabled?: boolean;
}

export interface DomainValidationRequest {
  appId: string;
  domain: string;
}

export interface DomainValidationResponse {
  isValid: boolean;
  reason?: string;
  appConfig?: AppConfigOutput;
}

/**
 * 创建应用配置
 * @param appConfig 应用配置信息
 * @returns 创建结果
 */
export async function createAppConfig(appConfig: AppConfigInput): Promise<ApiResponse<AppConfigOutput>> {
  return fetchApi<AppConfigOutput>(`${API_URL}/api/AppConfig`, {
    method: 'POST',
    body: JSON.stringify(appConfig),
  });
}

/**
 * 获取应用配置列表
 * @returns 应用配置列表
 */
export async function getAppConfigs(): Promise<ApiResponse<AppConfigOutput[]>> {
  return fetchApi<AppConfigOutput[]>(`${API_URL}/api/AppConfig`);
}

/**
 * 根据 AppId 获取应用配置
 * @param appId 应用ID
 * @returns 应用配置
 */
export async function getAppConfigByAppId(appId: string): Promise<ApiResponse<AppConfigOutput>> {
  return fetchApi<AppConfigOutput>(`${API_URL}/api/AppConfig/${encodeURIComponent(appId)}`);
}

/**
 * 更新应用配置
 * @param appId 应用ID
 * @param appConfig 应用配置信息
 * @returns 更新结果
 */
export async function updateAppConfig(appId: string, appConfig: AppConfigInput): Promise<ApiResponse<AppConfigOutput>> {
  return fetchApi<AppConfigOutput>(`${API_URL}/api/AppConfig/${encodeURIComponent(appId)}`, {
    method: 'PUT',
    body: JSON.stringify(appConfig),
  });
}

/**
 * 删除应用配置
 * @param appId 应用ID
 * @returns 删除结果
 */
export async function deleteAppConfig(appId: string): Promise<ApiResponse<void>> {
  return fetchApi<void>(`${API_URL}/api/AppConfig/${encodeURIComponent(appId)}`, {
    method: 'DELETE',
  });
}

/**
 * 启用/禁用应用
 * @param appId 应用ID
 * @returns 切换结果
 */
export async function toggleAppConfigEnabled(appId: string): Promise<ApiResponse<AppConfigOutput>> {
  return fetchApi<AppConfigOutput>(`${API_URL}/api/AppConfig/${encodeURIComponent(appId)}/toggle`, {
    method: 'POST',
  });
}

/**
 * 域名验证
 * @param request 验证请求
 * @returns 验证结果
 */
export async function validateDomain(request: DomainValidationRequest): Promise<ApiResponse<DomainValidationResponse>> {
  return fetchApi<DomainValidationResponse>(`${API_URL}/api/AppConfig/validatedomain`, {
    method: 'POST',
    body: JSON.stringify(request),
  });
}

/**
 * 获取公开应用配置
 * @param appId 应用ID
 * @returns 公开应用配置
 */
export async function getPublicAppConfig(appId: string): Promise<ApiResponse<AppConfigOutput>> {
  return fetchApi<AppConfigOutput>(`${API_URL}/api/AppConfig/public/${encodeURIComponent(appId)}`);
}

/**
 * 生成随机应用ID
 * @returns 随机应用ID
 */
export function generateAppId(): string {
  const timestamp = Date.now().toString(36);
  const randomStr = Math.random().toString(36).substring(2, 8);
  return `app_${timestamp}_${randomStr}`;
}

/**
 * 获取仓库列表（用于选择器）
 * @returns 仓库列表
 */
export async function getRepositoriesForSelector(): Promise<ApiResponse<{ organizationName: string; name: string; }[]>> {
  try {
    // 这里可以调用现有的仓库获取 API
    // 暂时返回空数组，可以根据实际需要修改
    return {
      code: 200,
      data: [],
      message: 'Success'
    };
  } catch (error) {
    return {
      code: 500,
      data: [],
      message: error instanceof Error ? error.message : '获取仓库列表失败'
    };
  }
} 

File: src/KoalaWiki/Services/AppConfigService.cs
using System.Text.Json;
using FastService;
using KoalaWiki.Domains;
using KoalaWiki.Dto;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace KoalaWiki.Services;

[Tags("应用配置管理")]
[FastService.Route("/api/AppConfig")]
[Filter(typeof(ResultFilter))]
public class AppConfigService(IKoalaWikiContext koala, IUserContext userContext) : FastApi
{
    /// <summary>
    /// 创建应用配置
    /// </summary>
    [FastService.Route("/")]
    public async Task<AppConfigOutput> CreateAsync(AppConfigInput input)
    {
        // 检查 AppId 是否已存在
        var existingApp = await koala.AppConfigs
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.AppId == input.AppId);

        if (existingApp != null)
        {
            throw new Exception("应用ID已存在");
        }

        // 检查用户是否有权限访问该组织和仓库
        var warehouse = await koala.Warehouses
            .AsNoTracking()
            .FirstOrDefaultAsync(x =>
                x.OrganizationName.ToLower() == input.OrganizationName.ToLower() &&
                x.Name.ToLower() == input.RepositoryName.ToLower());

        if (warehouse == null)
        {
            throw new Exception("指定的组织或仓库不存在");
        }

        var appConfig = new AppConfig
        {
            Id = Guid.NewGuid().ToString(),
            AppId = input.AppId,
            Name = input.Name,
            OrganizationName = input.OrganizationName,
            RepositoryName = input.RepositoryName,
            AllowedDomainsJson = JsonSerializer.Serialize(input.AllowedDomains),
            EnableDomainValidation = input.EnableDomainValidation,
            Description = input.Description,
            UserId = userContext.CurrentUserId ?? throw new Exception("用户未登录"),
            IsEnabled = true,
            CreatedAt = DateTime.UtcNow
        };

        await koala.AppConfigs.AddAsync(appConfig);
        await koala.SaveChangesAsync();

        return MapToOutput(appConfig);
    }

    /// <summary>
    /// 获取应用配置列表
    /// </summary>
    [FastService.Route("/")]
    public async Task<List<AppConfigOutput>> GetAsync()
    {
        var appConfigs = await koala.AppConfigs
            .AsNoTracking()
            .Where(x => x.UserId == userContext.CurrentUserId)
            .OrderByDescending(x => x.CreatedAt)
            .ToListAsync();

        return appConfigs.Select(MapToOutput).ToList();
    }

    /// <summary>
    /// 根据 AppId 获取应用配置
    /// </summary>
    [HttpGet("/{appId}")]
    public async Task<AppConfigOutput> GetByAppIdAsync(string appId)
    {
        var appConfig = await koala.AppConfigs
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.AppId == appId && x.UserId == userContext.CurrentUserId);

        if (appConfig == null)
        {
            throw new Exception("应用配置不存在");
        }

        return MapToOutput(appConfig);
    }

    /// <summary>
    /// 更新应用配置
    /// </summary>
    [HttpPut("/{appId}")]
    public async Task<AppConfigOutput> UpdateAsync(string appId, AppConfigInput input)
    {
        var appConfig = await koala.AppConfigs
            .FirstOrDefaultAsync(x => x.AppId == appId && x.UserId == userContext.CurrentUserId);

        if (appConfig == null)
        {
            throw new Exception("应用配置不存在");
        }

        // 检查新的组织和仓库是否存在（如果有变更）
        if (appConfig.OrganizationName != input.OrganizationName ||
            appConfig.RepositoryName != input.RepositoryName)
        {
            var warehouse = await koala.Warehouses
                .AsNoTracking()
                .FirstOrDefaultAsync(x =>
                    x.OrganizationName.ToLower() == input.OrganizationName.ToLower() &&
                    x.Name.ToLower() == input.RepositoryName.ToLower());

            if (warehouse == null)
            {
                throw new Exception("指定的组织或仓库不存在");
            }
        }

        appConfig.Name = input.Name;
        appConfig.OrganizationName = input.OrganizationName;
        appConfig.RepositoryName = input.RepositoryName;
        appConfig.AllowedDomainsJson = JsonSerializer.Serialize(input.AllowedDomains);
        appConfig.EnableDomainValidation = input.EnableDomainValidation;
        appConfig.Description = input.Description;

        await koala.SaveChangesAsync();

        return MapToOutput(appConfig);
    }

    /// <summary>
    /// 删除应用配置
    /// </summary>
    [HttpDelete("/{appId}")]
    public async Task DeleteAsync(string appId)
    {
        var appConfig = await koala.AppConfigs
            .FirstOrDefaultAsync(x => x.AppId == appId && x.UserId == userContext.CurrentUserId);

        if (appConfig == null)
        {
            throw new Exception("应用配置不存在");
        }

        koala.AppConfigs.Remove(appConfig);
        await koala.SaveChangesAsync();
    }

    /// <summary>
    /// 启用/禁用应用
    /// </summary>
    [HttpPost("/{appId}/toggle")]
    public async Task<AppConfigOutput> ToggleEnabledAsync(string appId)
    {
        var appConfig = await koala.AppConfigs
            .FirstOrDefaultAsync(x => x.AppId == appId && x.UserId == userContext.CurrentUserId);

        if (appConfig == null)
        {
            throw new Exception("应用配置不存在");
        }

        await koala.AppConfigs.Where(x => x.AppId == appId && x.UserId == userContext.CurrentUserId)
            .ExecuteUpdateAsync(x => x.SetProperty(a => a.IsEnabled, a => !a.IsEnabled));

        return MapToOutput(appConfig);
    }

    /// <summary>
    /// 域名验证（公开接口，不需要登录）
    /// </summary>
    [HttpPost("/validatedomain")]
    [AllowAnonymous]
    public async Task<DomainValidationResponse> ValidateDomainAsync(DomainValidationRequest request)
    {
        var appConfig = await koala.AppConfigs
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.AppId == request.AppId && x.IsEnabled);

        if (appConfig == null)
        {
            return new DomainValidationResponse
            {
                IsValid = false,
                Reason = "应用配置不存在或已禁用"
            };
        }

        // 如果未启用域名验证，直接通过
        if (!appConfig.EnableDomainValidation)
        {
            // 更新最后使用时间
            await UpdateLastUsedTimeAsync(appConfig.Id);

            return new DomainValidationResponse
            {
                IsValid = true,
                AppConfig = MapToOutput(appConfig)
            };
        }

        // 解析允许的域名列表
        var allowedDomains = JsonSerializer.Deserialize<List<string>>(appConfig.AllowedDomainsJson) ??
                             new List<string>();

        if (allowedDomains.Count == 0)
        {
            return new DomainValidationResponse
            {
                IsValid = false,
                Reason = "未配置允许的域名"
            };
        }

        // 验证域名
        var isValidDomain = allowedDomains.Any(domain =>
            request.Domain.Equals(domain, StringComparison.OrdinalIgnoreCase) ||
            request.Domain.EndsWith("." + domain, StringComparison.OrdinalIgnoreCase));

        if (!isValidDomain)
        {
            return new DomainValidationResponse
            {
                IsValid = false,
                Reason = $"域名 {request.Domain} 未被授权"
            };
        }

        // 更新最后使用时间
        await UpdateLastUsedTimeAsync(appConfig.Id);

        return new DomainValidationResponse
        {
            IsValid = true,
            AppConfig = MapToOutput(appConfig)
        };
    }

    /// <summary>
    /// 获取应用配置（公开接口，用于第三方脚本）
    /// </summary>
    [HttpGet("/public/{appId}")]
    [AllowAnonymous]
    public async Task<AppConfigOutput?> GetPublicConfigAsync(string appId)
    {
        var appConfig = await koala.AppConfigs
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.AppId == appId && x.IsEnabled);

        if (appConfig == null)
        {
            return null;
        }

        // 更新最后使用时间
        await UpdateLastUsedTimeAsync(appConfig.Id);

        return MapToOutput(appConfig);
    }

    /// <summary>
    /// 更新最后使用时间
    /// </summary>
    private async Task UpdateLastUsedTimeAsync(string appConfigId)
    {
        await koala.AppConfigs
            .Where(x => x.Id == appConfigId)
            .ExecuteUpdateAsync(x => x.SetProperty(p => p.LastUsedAt, DateTime.UtcNow));
    }

    /// <summary>
    /// 将实体映射为输出 DTO
    /// </summary>
    private static AppConfigOutput MapToOutput(AppConfig appConfig)
    {
        var allowedDomains = JsonSerializer.Deserialize<List<string>>(appConfig.AllowedDomainsJson) ??
                             new List<string>();

        return new AppConfigOutput
        {
            AppId = appConfig.AppId,
            Name = appConfig.Name,
            OrganizationName = appConfig.OrganizationName,
            RepositoryName = appConfig.RepositoryName,
            AllowedDomains = allowedDomains,
            EnableDomainValidation = appConfig.EnableDomainValidation,
            Description = appConfig.Description,
            IsEnabled = appConfig.IsEnabled,
            CreatedAt = appConfig.CreatedAt,
            UpdatedAt = appConfig.CreatedAt // 这里可以根据需要添加 UpdatedAt 字段
        };
    }
}